src/drivers/rpi_pca9685_pwm_out/PCA9685.cpp

```diff
@@ -0,0 +1,229 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ *
+ * Original Author      : Georgi Todorov
+ * Edited by	: Tord Wessman
+ * Created on  : Nov 22, 2013
+ * Rewrite	   : Fan.zhang  421395590@qq.com
+ * Updated on  : Mar 2, 2017
+ ****************************************************************************/
+
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <linux/i2c-dev.h>
+#include <stdio.h>      /* Standard I/O functions */
+#include <fcntl.h>
+#include <syslog.h>		/* Syslog functionallity */
+#include <inttypes.h>
+#include <errno.h>
+#include <math.h>
+#include <stdio.h>
+#include <unistd.h>
+//
+
+#include "PCA9685.h"
+
+//! Constructor takes bus and address arguments
+/*!
+ \param bus the bus to use in /dev/i2c-%d.
+ \param address the device address on bus
+ */
+
+void PCA9685::init(int bus, int address)
+{
+	_i2cbus = bus;
+	_i2caddr = address;
+	snprintf(busfile, sizeof(busfile), "/dev/i2c-%d", bus);
+	reset();
+	//usleep(10*1000);
+}
+
+
+PCA9685::PCA9685() :
+	_i2caddr(PCA9685_DEFAULT_I2C_ADDR),
+	_i2cbus(PCA9685_DEFAULT_I2C_BUS),
+	dataBuffer {}
+{
+
+}
+
+PCA9685::PCA9685(int bus, int address) :
+	busfile {},
+	dataBuffer {}
+{
+	_i2cbus = bus;
+	_i2caddr = address;
+	snprintf(busfile, sizeof(busfile), "/dev/i2c-%d", bus);
+	reset();
+}
+
+PCA9685::~PCA9685()
+{
+	reset();
+}
+//! Sets PCA9685 mode to 00
+void PCA9685::reset()
+{
+	int fd = openfd();
+
+	if (fd != -1) {
+		write_byte(fd, MODE1, 0x00); //Normal mode
+		write_byte(fd, MODE2, 0x04); //Normal mode
+		close(fd);
+	}
+}
+//! Set the frequency of PWM
+/*!
+ \param freq desired frequency. 40Hz to 1000Hz using internal 25MHz oscillator.
+ */
+void PCA9685::setPWMFreq(int freq)
+{
+	int fd = openfd();
+
+	if (fd != -1) {
+		uint8_t prescale = (CLOCK_FREQ / MAX_PWM_RES / freq)  - 1;
+		//printf ("Setting prescale value to: %d\n", prescale);
+		//printf ("Using Frequency: %d\n", freq);
+
+		uint8_t oldmode = read_byte(fd, MODE1);
+		uint8_t newmode = (oldmode & 0x7F) | 0x10;    //sleep
+		write_byte(fd, MODE1, newmode);        // go to sleep
+		write_byte(fd, PRE_SCALE, prescale);
+		write_byte(fd, MODE1, oldmode);
+		usleep(10 * 1000);
+		write_byte(fd, MODE1, oldmode | 0x80);
+
+		close(fd);
+	}
+}
+
+//! PWM a single channel
+/*!
+ \param led channel to set PWM value for
```

Comment:
LED channel megs pwm Channel.
In china,pca9685 was known as LED controller.
I'll change these comment